### PR TITLE
feat: add btc-listener

### PR DESCRIPTION
**Pull request by tcar121293**
added btc listener
## Description
added btc listener and message handler
## Related Issue Or Context

Closes: #264 

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Documentation

## Checklist:

- [ ] I have commented my code, particularly in hard-to-understand areas.
- [ ] I have ensured that all acceptance criteria (or expected behavior) from issue are met
- [ ] I have updated the documentation locally and in docs.
- [ ] I have added tests to cover my changes.
- [ ] I have ensured that all the checks are passing and green, I've signed the CLA bot

### CODE REVIEW

File: chains/btc/listener/util.go [hunk 1]
```diff
@@ -0,0 +1,73 @@
...
+		if bridge == vout.ScriptPubKey.Address {
+			isBridgeDeposit = true
+
+			if vout.ScriptPubKey.Type == PubKeyHash || vout.ScriptPubKey.Type == ScriptHash || vout.ScriptPubKey.Type == WitnessV0KeyHash {
+				amount.Add(amount, big.NewInt(int64(vout.Value*1e8)))
+			}
+		}
+	}
+
+	if isBridgeDeposit {
```
Comment (mpetrun5) [inline chains/btc/listener/util.go:48]: Can you rewrite this as !isBridgeDeposit and return immediately if it isn't?
It is always more clearer to have the optimistic path last.

File: chains/btc/listener/util.go [hunk 2]
```diff
@@ -0,0 +1,73 @@
...
+		for _, vin := range evt.Vin {
```
Comment (mpetrun5) [inline chains/btc/listener/util.go:49]: Do we need this?
I don't think we really care who sent it tbh.

Comment (mpetrun5) [reply chains/btc/listener/util.go:49]: Correct if I'm wrong @MakMuftic @nmlinaric just to be sure.

File: chains/btc/listener/util.go [hunk 3]
```diff
@@ -0,0 +1,73 @@
...
+	"encoding/hex"
+	"math/big"
+
+	"github.com/ChainSafe/sygma-relayer/e2e/evm"
+	"github.com/btcsuite/btcd/btcjson"
+	"github.com/btcsuite/btcd/chaincfg/chainhash"
+	"github.com/ethereum/go-ethereum/common"
+)
+
+var RESOURCE_ID = evm.SliceTo32Bytes(common.LeftPadBytes([]byte{0}, 31))
```
Comment (mpetrun5) [inline chains/btc/listener/util.go:13]: I would prefer this to be config based so each event handler per address has configurable resourceID.
So:
 - address A --> resource 0x...01
 - address B --> resource 0x....02
 
And do that in the event handler

File: chains/btc/listener/event-handlers.go [hunk 4]
```diff
@@ -0,0 +1,142 @@
...
+			defer func() {
+				if r := recover(); r != nil {
+					log.Error().Msgf("panic occured while handling deposit %+v", evt)
+				}
+			}()
+
+			d, isDeposit, err := DecodeDepositEvent(evt, eh.conn, eh.bridge)
+			if err != nil {
+				log.Error().Err(err).Msgf("%v", err)
+				return
```
Comment (mpetrun5) [inline chains/btc/listener/event-handlers.go:76]: Would be nice to return these errors.
If you make the func(evt btcjson.TxRawResult) be func(evt btcjson.TxRawResult) error
Then you can just return from this function

File: chains/btc/listener/deposit-handler.go [hunk 5]
```diff
@@ -0,0 +1,54 @@
...
+
+	// data is composed of recieverEVMAddress_destinationDomainID
+	parsedData := strings.Split(data, "_")
+	evmAdd := common.HexToAddress(parsedData[0]).Bytes()
+	destDomainID, err := strconv.ParseUint(parsedData[1], 10, 8)
+	if err != nil {
+		return nil, err
+	}
+
+	payload := []interface{}{
```
Comment (mpetrun5) [inline chains/btc/listener/deposit-handler.go:41]: Create a type for this payload that can be reused on destination. 
I think a TransferMessageData probably exist.

File: chains/btc/connection/connection.go [hunk 6]
```diff
@@ -0,0 +1,31 @@
...
+)
+
+type Connection struct {
+	*rpcclient.Client
+}
+
+func NewBtcConnection(url string) (*Connection, error) {
+
+	// Connect to a Bitcoin node using RPC
+	connConfig := &rpcclient.ConnConfig{
```
Comment (mpetrun5) [inline chains/btc/connection/connection.go:17]: Don't we need the username and password here?

File: chains/btc/chain.go [hunk 7]
```diff
@@ -0,0 +1,94 @@
...
+}
+
+func (c *BtcChain) Write(msgs []*proposal.Proposal) error {
+	return nil
+}
+
+// remove this after
+func (c *BtcChain) ReceiveMessage(m *message.Message)/* *proposal.Proposal, error */ {}
+
+func (c *BtcChain) PollEvents(ctx context.Context) {
```
Comment (mpetrun5) [inline chains/btc/chain.go:53]: I think is now changed to accept start block so this logic should be moved to the app.

Comment (mpetrun5) [reply chains/btc/chain.go:53]: Look at EVM chain.go

File: chains/btc/listener/event-handlers.go [hunk 8]
```diff
@@ -0,0 +1,142 @@
...
+
+	// Fetch block details in verbose mode
+	block, err := eh.conn.GetBlockVerboseTx(blockHash)
+	if err != nil {
+		return nil, err
+	}
+	return block.Tx, nil
+}
+
+func (eh *FungibleTransferEventHandler) GetNonce(blockNumber *big.Int, evtNumber int) (uint64, error) {
```
Comment (mpetrun5) [inline chains/btc/listener/event-handlers.go:121]:
```suggestion
func (eh *FungibleTransferEventHandler) Nonce(blockNumber *big.Int, evtNumber int) (uint64, error) {
```

```suggestion
func (eh *FungibleTransferEventHandler) CalculateNonce(blockNumber *big.Int, evtNumber int) (uint64, error) {
```

Either of these two

**Review by mpetrun5 — CHANGES_REQUESTED** [review]

File: chains/btc/listener/deposit-handler.go (hunk 5 above)
Comment (tcar121293) [reply chains/btc/listener/deposit-handler.go:41]: I am using TransferMessageData, but payload is always just an interface, because the payload can be any data, depending on what kind of transfer we are doing. Erc20, GMP etc... then we parse the payload in the executor depending on the transfer type, right?

Comment (mpetrun5) [reply chains/btc/listener/deposit-handler.go:41]: Leave the type here as TransferMessageData because then we can just cast in the executor which is easier.
I think if you are using it as []interface here we cannot cast the type on the executor

File: chains/btc/config_test.go [hunk 9]
```diff
@@ -0,0 +1,83 @@
...
+		"endpoint":           "ws://domain.com",
+		"name":               "btc1",
+		"blockConfirmations": -1,
+	})
+
+	s.NotNil(err)
+	s.Equal(err.Error(), "blockConfirmations has to be >=1")
+}
+
+func (s *NewBtcConfigTestSuite) Test_ValidConfig() {
```
Comment (mpetrun5) [inline chains/btc/config_test.go:60]: Can you test all the other params?
Username, password...
I think they might not be propagated

File: chains/btc/connection/connection.go [hunk 10]
```diff
@@ -0,0 +1,33 @@
...
+import (
+	"github.com/btcsuite/btcd/rpcclient"
+)
+
+type Connection struct {
+	*rpcclient.Client
+}
+
+func NewBtcConnection(url string, username string, password string) (*Connection, error) {
+
```
Comment (mpetrun5) [inline chains/btc/connection/connection.go:15]:
```suggestion
```

File: chains/btc/config_test.go (hunk 9 above)
Comment (mpetrun5) [reply chains/btc/config_test.go:60]: Also username and password should probably be required

File: chains/btc/listener/deposit-handler.go [hunk 11]
```diff
@@ -0,0 +1,54 @@
...
+}
+
+func (e *BtcDepositHandler) HandleDeposit(sourceID uint8,
+	depositNonce uint64,
+	resourceID [32]byte,
+	amount *big.Int,
+	data string,
+	blockNumber *big.Int,
+) (*message.Message, error) {
+
```
Comment (mpetrun5) [inline chains/btc/listener/deposit-handler.go:32]:
```suggestion
```

File: chains/btc/listener/deposit-handler_test.go [hunk 12]
```diff
@@ -0,0 +1,89 @@
...
+type Erc20HandlerTestSuite struct {
+	suite.Suite
+}
+
+func TestRunErc20HandlerTestSuite(t *testing.T) {
+	suite.Run(t, new(Erc20HandlerTestSuite))
+}
+
+func (s *Erc20HandlerTestSuite) Test_Erc20HandleEvent() {
+
```
Comment (mpetrun5) [inline chains/btc/listener/deposit-handler_test.go:30]:
```suggestion
```

File: chains/btc/listener/event-handlers.go [hunk 13]
```diff
@@ -0,0 +1,144 @@
...
+			m, err := eh.depositHandler.HandleDeposit(eh.domainID, nonce, d.ResourceID, d.Amount, d.Data, blockNumber)
+			if err != nil {
+				return err
+			}
+
+			log.Debug().Str("messageID", m.ID).Msgf("Resolved message %+v in block: %s", m, blockNumber.String())
+			domainDeposits[m.Destination] = append(domainDeposits[m.Destination], m)
+			return nil
+		}(evt)
+
```
Comment (mpetrun5) [inline chains/btc/listener/event-handlers.go:95]:
```suggestion
```

File: chains/btc/listener/event-handlers.go [hunk 14]
```diff
@@ -0,0 +1,144 @@
...
+	// Fetch block details in verbose mode
+	block, err := eh.conn.GetBlockVerboseTx(blockHash)
+	if err != nil {
+		return nil, err
+	}
+	return block.Tx, nil
+}
+
+func (eh *FungibleTransferEventHandler) CalculateNonce(blockNumber *big.Int, evtNumber int) (uint64, error) {
+
```
Comment (mpetrun5) [inline chains/btc/listener/event-handlers.go:124]:
```suggestion
```

File: chains/btc/listener/listener.go [hunk 15]
```diff
@@ -0,0 +1,109 @@
...
+			}
+
+			// Sleep if startBlock is higher then head
+			if new(big.Int).Sub(head, startBlock).Cmp(l.blockConfirmations) == -1 {
+				time.Sleep(l.blockRetryInterval)
+				continue
+			}
+
+			for _, handler := range l.eventHandlers {
+				err := handler.HandleEvents(startBlock)
```
Comment (mpetrun5) [inline chains/btc/listener/listener.go:93]: We should probably make this a range startBlock -- endBlock (same as substrate), but let's create an issue for this and do it in a separate PR after everything is working on devent

File: chains/btc/listener/util.go [hunk 16]
```diff
@@ -0,0 +1,71 @@
...
+			}
+		}
+	}
+
+	if !isBridgeDeposit {
+		return Deposit{}, false, nil
+	}
+
+	for _, vin := range evt.Vin {
+
```
Comment (mpetrun5) [inline chains/btc/listener/util.go:50]:
```suggestion
```

**Review by mpetrun5 — CHANGES_REQUESTED** [review]

File: chains/btc/config.go [hunk 17]
```diff
@@ -0,0 +1,85 @@
...
+		return err
+	}
+
+	if c.BlockConfirmations != 0 && c.BlockConfirmations < 1 {
+		return fmt.Errorf("blockConfirmations has to be >=1")
+	}
+	return nil
+}
+
+type ResourceConfig struct {
```
Comment (mpetrun5) [inline chains/btc/config.go:38]:
```suggestion
type Resource struct {
```

**Review by mpetrun5 — CHANGES_REQUESTED** [review]

File: chains/btc/listener/util.go (hunk 2 above)
Comment (nmlinaric) [reply chains/btc/listener/util.go]: I would agree here, `sender` is used only to lock/burn tokens and charge fee programmatically in Solidity, now this will be done in taproot scripts

**Review by nmlinaric — APPROVED** [review]

**Review by MakMuftic — APPROVED** [review]
